File: app/src/main/java/com/practice/todo/util/RemindUtil.java
package com.practice.todo.util;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.text.TextUtils;

import androidx.core.app.NotificationCompat;

import com.practice.todo.App;
import com.practice.todo.MainActivity;
import com.practice.todo.R;
import com.practice.todo.storage.database.entity.TodoItem;

/**
 * 提醒工具类
 *
 * 这个app中的提醒是一个通知加上震动，适配Android8.0需要加上Channel。
 */
public class RemindUtil {

    private static final String NOTIFICATION_CHANNEL_ID = "10051";
    private static final String default_notification_channel_id = "todo_practice2";

    public static void remind(TodoItem todoItem) {
        Intent ntfIntent = new Intent(App.instance, MainActivity.class);
        ntfIntent.putExtra("fromNotification", true);
        ntfIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent pendingIntent = PendingIntent.getActivity(App.instance, 0, ntfIntent, 0);
        NotificationManager nm = (NotificationManager) App.instance.getSystemService(Context.NOTIFICATION_SERVICE);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(App.instance, default_notification_channel_id);

        final String contentText = TextUtils.isEmpty(todoItem.getDescription()) ?
                "You can mark this todo is done" :
                todoItem.getDescription();

        builder.setContentTitle(todoItem.getTitle());
        builder.setContentIntent(pendingIntent);
        builder.setContentText(contentText);
        builder.setSmallIcon(R.drawable.ic_launcher);
        builder.setAutoCancel(true);

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            int importance = NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel channel = new NotificationChannel(
                    NOTIFICATION_CHANNEL_ID,
                    "NOTIFICATION_CHANNEL_NAME", importance);

            channel.enableVibration(true);
            channel.canBypassDnd();
            channel.setLockscreenVisibility(Notification.VISIBILITY_SECRET);
            channel.enableLights(true);
            channel.setLightColor(Color.RED);
            channel.setBypassDnd(true);
            channel.shouldShowLights();

            builder.setChannelId(NOTIFICATION_CHANNEL_ID);
            nm.createNotificationChannel(channel);
        }
        nm.notify((int) System.currentTimeMillis(), builder.build());
    }
}

File: app/src/main/java/com/practice/todo/util/RemindUtil.kt
package com.practice.todo.util

import android.app.Notification.VISIBILITY_SECRET
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Color
import androidx.core.app.NotificationCompat
import com.practice.todo.App
import com.practice.todo.MainActivity
import com.practice.todo.R
import com.practice.todo.storage.database.entity.TodoItem

/**
 * Created by suikajy on 2019.10.2
 */
object NotificationUtil {

    private const val NOTIFICATION_CHANNEL_ID = "10051"
    private const val default_notification_channel_id = "todo_practice2"

    fun notification(todoItem: TodoItem) {
        val notificationIntent = Intent(App.instance, MainActivity::class.java)
        notificationIntent.putExtra("fromNotification", true)
        notificationIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        val pendingIntent = PendingIntent.getActivity(App.instance, 0, notificationIntent, 0)
        val mNotificationManager =
            App.instance.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val mBuilder =
            NotificationCompat.Builder(
                App.instance,
                default_notification_channel_id
            )
        mBuilder.setContentTitle(todoItem.title)
            .setContentIntent(pendingIntent)
            .setContentText("You can mark this todo is done")
            .setSmallIcon(R.drawable.ic_launcher)
            .setAutoCancel(true)
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_HIGH
            val channel = NotificationChannel(
                NOTIFICATION_CHANNEL_ID,
                "NOTIFICATION_CHANNEL_NAME",
                importance
            )
            channel.enableVibration(true)
            channel.canBypassDnd()
            channel.lockscreenVisibility = VISIBILITY_SECRET
            channel.enableLights(true)
            channel.lightColor = Color.RED
            channel.audioAttributes
            channel.group
            channel.setBypassDnd(true)
            channel.shouldShowLights()

            mBuilder.setChannelId(NOTIFICATION_CHANNEL_ID)
            mNotificationManager.createNotificationChannel(channel)
        }
        mNotificationManager.notify(System.currentTimeMillis().toInt(), mBuilder.build())
    }

}